docs/proposals/14-reversed-cluster-vpn.md

```diff
@@ -0,0 +1,276 @@
+
+duplicate-cn
+
+key "/srv/secrets/vpn-shoot/tls.key"
+cert "/srv/secrets/vpn-shoot/tls.crt"
+ca "/srv/secrets/vpn-shoot/ca.crt"
+dh "/srv/secrets/dh/dh2048.pem"
+
+tls-auth "/srv/secrets/tlsauth/vpn.tlsauth" 0
+push "route 10.242.0.0 255.255.0.0"
+```
+
+It is a TCP TLS server and configured to automatically assign IP addresses for OpenVPN clients (`server` directive). In addition, it pushes the shoot cluster node-, pod-, and service ranges to the clients running in the seed cluster (`push` directive). 
+
+**Note:** The network mesh spanned by OpenVPN uses the network range `192.168.123.0 - 192.168.123.255`. This network range cannot be used in either shoot-, or seed clusters. If it is used this might cause subtle problem due to network range overlaps. Unfortunately, this appears not to be well documented but this restriction exists since the very beginning. We should clean up this technical debt as part of the excercise.
+
+### Goals
+
+* We intend to supersede the current VPN solution with the solution outlined in this proposal.
+* We intend to remove the code for the konnectivity tunnel once this solution proposal has been validated.
+
+### Non Goals
+
+* The solution is not a low latency, or high throughput solution. As the kube-apiserver to shoot cluster traffic does not demand these properties we do not intend to invest in improvements.
+* We do not intend to provide continuous availability to the shoot-seed VPN connection. We expect the availability to be comparable to the existing solution.
+
+## Proposal
+
+The propsal is depicted in the following diagram:
+
+![alt text](assets/ReversedTunnelVPN.png "Reversed Tunnel VPN")
+
+We have added an OpenVPN server pod (`vpn-seed-server`) to each control plane. This OpenVPN server is configured with the OpenVPN `client-to-client` option that allows clients to talk to each other. The OpenVPN client in the shoot cluster (`vpn-shoot-client`) as well as the Open VPN client sidecars in the seed (`vpn-seed-client`) connect to the OpenVPN server.
+
+The three containers `vpn-seed-server`, `vpn-seed-client`, and `vpn-shoot-client` are new containers and are not related to containers in the github.com/gardener/vpn project. We will create a new project github.com/gardener/vpn2 for these containers. With this solution we intend to supersede the containers from the github.com/gardener/vpn project.
+
+A service `openvpn-server` is created for each control plane in its namespace. The `vpn-seed-client` sidecars connect directly to this service.
+
+The `vpn-shoot-client` pod connects to the correct `vpn-seed-server` service via the SNI passthrough proxy introduced with [SNI Passthrough proxy for kube-apiservers](https://github.com/gardener/gardener/blob/master/docs/proposals/08-shoot-apiserver-via-sni.md) port 443. No additional port needs to be opened. We re-use the SNI infrastructure in order to avoid the need to have one external load balancer per shoot cluster in the seed.
```

Comment:
This is not correct - the `openvpn-client` should use `HTTP PROXY` (use `CONNECT` verb to open connection to the right `openvpn-server`). Setting the ServerName in TLS CLIENT HELLO support is only added in `openvpn` `v3` that we do not use. https://www.envoyproxy.io/docs/envoy/v1.17.1/intro/arch_overview/http/upgrades.html?highlight=upgrade_configs#connect-support

Reply:
You are correct that the `vpn-shoot-client` uses the envoy as HTTP proxy as SNI is only supported in `openvpn`v3. Dirk will adjust the sentence to be more precise.

Reply:
This is correctly stated in the next paragraph:

> Shoot OpenVPN clients connect to the correct OpenVPN Server using the http proxy feature provided by OpenVPN. A configuration is added to the envoy proxy to detect http proxy requests and forward the connection attempt to the correct OpenVPN server.